sql: properly implement create_type_statements index

otan opened the pull request:
See individual commits for details!

Resolves #71414

otan committed:
sql: properly implement create_type_statements index

Release note (sql change): The `create_type_statements` table now has an
index on `descriptor_id`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73635)
<!-- Reviewable:end -->

rafiss commented on pkg/sql/logictest/testdata/logic_test/crdb_internal:
```diff
@@ -760,6 +760,15 @@ SELECT * FROM crdb_internal.create_type_statements
+SELECT * FROM crdb_internal.create_type_statements WHERE descriptor_id = 'foo'::regclass::oid
+----
+
 # Test the virtual index as well.
```
my fav part of this is the pre-existing comment here :P

otan replied:
ooh moved this yp!

rafiss commented on pkg/sql/crdb_internal.go:
```diff
@@ -2243,53 +2291,60 @@ CREATE TABLE crdb_internal.create_type_statements (
+	},
+	indexes: []virtualIndex{
+		{
+			populate: func(
```
can this share code with the index on pg_catalog.pgtype?

otan replied:
ok

rafiss reviewed (approved):
thank you!

> Reviewed 3 of 4 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73635#-:-MqUYBXTCb2m3pQpanIY:bsv417h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

otan commented:
bors r-

otan committed:
sql: add index check for virtualTables

Check that the indexes in the schema for the virtualTable matches the
number of index array elements.

Release note: None

otan commented:
bors r=rafiss
